pybind11/how-to.rst

```diff
@@ -553,88 +631,146 @@ and ``ExampleThree.h``:
         T _value;
     };
 
-    }
-    }  // namespace lsst::tmpl
+    }}  // namespace lsst::tmpl
 
     #endif
 
-Create wrapper files
-^^^^^^^^^^^^^^^^^^^^
+More wrapper files
+------------------
 
-Again following :ref:`our rules on file naming <style-guide-pybind11-module-naming>` we create a basic file for the wrapper ``python/lsst/TMPL/exampleTwo.cc`` (note that this file will later move to a subpackage):
+Because code in different headers should usually be wrapped in different source files, we'll create two new skeletons for ``ExampleTwo.h`` and ``ExampleThree.h``:
 
 .. code-block:: cpp
 
+    // _ExampleTwo.cc
+
     #include "pybind11/pybind11.h"
 
-    #include "lsst/TMPL/ExampleTwo.h"
+    #include "lsst/utils/python.h"
+    #include "lsst/tmpl/ExampleTwo.h"
 
     namespace py = pybind11;
     using namespace pybind11::literals;
 
-    namespace lsst {
-    namespace tmpl {
+    namespace lsst { namespace tmpl {
+
+    void wrapExampleTwo(utils::python::WrapperCollection & wrappers) {
+        // ... wrappers for ExampleTwo go here ...
+    }
+
+    }}  // lsst::tmpl
+
+.. code-block:: cpp
+
+    // _ExampleThree.cc
 
-    PYBIND11_MODULE(exampleTwo, mod) {
+    #include "pybind11/pybind11.h"
+
+    #include "lsst/utils/python.h"
+    #include "lsst/tmpl/ExampleThree.h"
+
+    namespace py = pybind11;
+    using namespace pybind11::literals;
+
+    namespace lsst { namespace tmpl {
+
+    void wrapExampleThree(utils::python::WrapperCollection & wrappers) {
+        // ... wrappers for ExampleThree go here ...
     }
 
     }}  // lsst::tmpl
 
-And the same for ``exampleThree.cc``.
+
+Our main source file should then be updated to declare and call all of the ``wrap*`` functions:
+
+.. code-block:: cpp
+
+    // _tmpl.cc
+
+    #include "pybind11/pybind11.h"
+
+    #include "lsst/utils/python.h"
+
+    namespace lsst { namespace tmpl {
+
+    void wrapExampleOne(utils::python::WrapperCollection & wrappers);
```

Comment:
Maybe it's making things too magic, but I was wondering if you couldn't reduce boilerplate by defining a few preprocessor macros to make this:
```
HEADER(ExampleOne);
HEADER(ExampleTwo);
HEADER(ExampleThree);

PYBIND11_MODULE(_tmpl, mod) {
    WRAP_COLLECTION(mod, "lsst.tmpl"); // automatically gets list of headers from HEADER macros and includes finish
}
```

Reply:
My first thought is that it's too magical, though I could be persuaded.  This would probably work for most packages, but it might be confusing that the order of `HEADER`s can matter (for inheritance dependencies) if it sets the order in `WRAP_COLLECTION`.  It would not work for afw, where we do need binary submodules - it's unfortunately both a common example and a bad one for future code.

FWIW, I don't immediately see how you'd implement `WRAP_COLLECTION`'s getting the list from the `HEADER` invocations - maybe kind of static vector of function pointers?

Reply:
Yes, let's leave it as is for now.  (I _was_ thinking of a function pointer vector, although namespacing and initialization of that vector were going to be potential issues.)